### ISSUE TITLE
C++ api in Debug x64 mode not building for VS2015

### ISSUE THREAD

**Issue by pabloxrl** [tensorflow/tensorflow#10867]
Hi,

I'm not sure if this is expected behaviour or a problem with the solution file generated by CMake, so my apologies if I'm writing in the wrong place.

I'm trying to build Tensorflow C++ API on Windows using VS2015, and so far I didn't have problems to build Release x64 binaries, which just required me to ensure that the toolset for 64 bits is in use when creating the def file (dump bin.exe and undname.exe in particular).

I could verify that the build works as expected, being able to load models saved in Python and perform inference successfully.

However, when building the debug x64 version of the .dll, I'm being blocked by the following error:

```
104>  tensorflow_static.vcxproj -> D:\out\Debug\tensorflow_static.lib
104>  symbols=1016688, taken=140782, dupes=3472
105>------ Build started: Project: tensorflow, Configuration: Debug x64 ------
105>  Building Custom Rule D:/tensorflow/tensorflow/contrib/cmake/CMakeLists.txt
105>  CMake does not need to re-run because D:/out/CMakeFiles/generate.stamp is up-to-date.
105>LINK : fatal error LNK1189: library limit of 65535 objects exceeded
========== Build: 104 succeeded, 1 failed, 1 up-to-date, 0 skipped ==========
```

Is there any workaround for this issue? Has anyone managed to build the debug version of the library? Is it possible at all?

From what I could google around, this linker issue is not easy to overcome, so feedback is most welcome.

**Comment by tatatodd** [comment]
Perhaps @mrry or @asimshankar might have some ideas here.

**Comment by pabloxrl** [comment]
As a sidenote, I'm trying to build other configurations available in the solution like: MinSizeRel or RelWithDebug and get hit by compiler errors of the form:

`d:\out\external\eigen_archive\eigen\src/Core/products/GeneralMatrixVector.h(351): fatal error C1060: compiler is out of heap space (compiling source file D:\tensorflow\tensorflow\core\kernels\matmul_op.cc)`

**Comment by pabloxrl** [comment]
After digging for a while I found the following (https://github.com/tensorflow/tensorflow/blob/master/tensorflow/contrib/cmake/README.md):

> Note that the -DCMAKE_BUILD_TYPE=Release flag must match the build configuration that you choose when invoking msbuild. The known-good values are Release and RelWithDebInfo. The Debug build type is not currently supported, because it relies on a Debug library for Python (python35d.lib) that is not distributed by default.
> 
> There are various options that can be specified when generating the solution and project files:
> 
> -DCMAKE_BUILD_TYPE=(Release|RelWithDebInfo): Note that the CMAKE_BUILD_TYPE option must match the build configuration that you choose when invoking MSBuild in step 4. The known-good values are Release and RelWithDebInfo. The Debug build type is not currently supported, because it relies on a Debug library for Python (python35d.lib) that is not distributed by default.

Which is consistent with my observations. I will try with RelWithDebInfo on CMake and see what happens.

**Comment by mrlzla** [comment]
I suppose that if you need only C++ API then you need only python3.exe to run script that check versions and create_def_file.py. The problem is related to the second one script.

**Comment by pabloxrl** [comment]
@mrlzla What do you mean by second one script? 

My problem -just to rephrase- is that I could not manage to build the debug version of tensorflow.dll. As far as I see this is not supported... Does this mean that I cannot have a project built in debug configuration which uses tensorflow.dll in VS2015?

**Comment by mrlzla** [comment]
```create_def_file.py```.It seems like we have to reduce the number of symbols that is exported from tensorflow.dll file.
As we can see in ```create_def_file.py```:
```
Because the linker allows only 64K symbols to be exported per dll
we filter the symbols down to the essentials. The regular expressions
we use for this are specific to tensorflow.
```
the symbols are already filtered, but the number of them still more than 65535

**Comment by Garoe** [comment]
As a workaround compiling tensorflow_static.lib in debug should work.

The `compiler is out of heap space` error happens if you haven't activated the 64bit toolset. Note that the `vcvarsall.bat` command only affects the `MSBuild` command and if you want visual studio to use the 64bit toolset you need to launch it with this commands
`set PreferredToolArchitecture=x64`
`"C:\Program Files (x86)\Microsoft Visual Studio 14.0\Common7\IDE\devenv.exe"`

**Comment by gunan** [comment]
@guschmue as he looked into this specific issue

**Comment by guschmue** [comment]
I know the reason  why the debug build runs into the 65535 symbol linker limit:
The debug build does not expand inlines. The fix is to change it CMakeLists.txt:
set(CMAKE_CXX_FLAGS_DEBUG "/D_DEBUG /MDd /Ob0")
to
set(CMAKE_CXX_FLAGS_DEBUG "/D_DEBUG /MDd /Ob2")

I'll send a PR for this.

**Comment by pabloxrl** [comment]
@Garoe I have made sure that the 64 bit toolset is being used and while I can build Release, VS2015 still fails to build the other configurations due to ` compiler is out of heap space`.

@guschmue Thank you very much for pointing out this issue. I tried changing CMakeLists.txt but now Debug build crashes earlier due to not enough heap space for the compiler:

`57>d:\tensorflow-1.2.0\tensorflow\contrib\cmake\build\external\eigen_archive\eigen\src\core\products\generalblockpanelkernel.h(1902): fatal error C1002: compiler is out of heap space in pass 2`

I'm not sure if this fix that you propose will not uncover different issues.

**Comment by guschmue** [comment]
I have seen this on a 8GB box in the past. msvc uses lots memory for some of files in tf, believed to be related to the amount of templates used in tf. I know there are bugs filed for msvc and hope this will improve in the future.
In the meantime you could try the following:
- kill other apps that use a lot of memory (in my case I had pycharm running in the background)
- check the pagefile size and give it 16 or 24GB

**Comment by haiy** [comment]
hi @pabloxrl, how do you build the tf with vs? Would you please give some detail about the process?I'm trying to debug the code with an IDE on Mac,but did't know how to do it.Do you have any blog or gist?

**Comment by pabloxrl** [comment]
@haiy It is worth noting that I could only manage so far to build Release x64 using VS2015. Building Debug x64 doesn't seem problematic according to @guschmue remarks, but I have space limitations in my machine so it will take me few days to verify that.

For Release x64 I followed the steps specified in https://github.com/tensorflow/tensorflow/blob/master/tensorflow/contrib/cmake/README.md, with the exception that I didn't enable CUDA GPU support. I encourage you to use the recommended setup as it is the less-painful way to get started. Nevertheless, the whole page is worth a read (see, current known limitations section).

The only special thing to add, if the C++ API is what you're after, would be to enable the flag: `-Dtensorflow_BUILD_SHARED_LIB` when you run cmake, which is what on Windows gives you the option to build `tensorflow_static.lib`, `tensorflow.lib` and `tensorflow.dll`.

I'll try to drop by to comment on my progress with the Debug build.

**Comment by BowieHsu** [comment]
https://github.com/BowieHsu/Tensorflow-windows-dll wish this could help you, complied tensorflow gpu dll

**Comment by tensorflowbutler** [comment]
It has been 14 days with no activity and the `awaiting tensorflower` label was assigned. Please update the label and/or status accordingly.

**Comment by dksb** [comment]
Closing due to lack of recent activity. If you are still facing this problem, please create a new issue using one of the issue templates. Thanks!
